### PR TITLE
Basic Testing Framework

### PR DESCRIPTION
**Pull request by aletya**
Notes:
* Changed connecting to database to a middleware as a workaround for jest
* basic testing works
* 
* To-do in next iteration: 
* implement a test suite for a service to make sure it works
* fix enum usage: https://github.com/colinhacks/zod/discussions/2125
* add instructions on how to make tests

### CODE REVIEW

File: src/app.test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+describe("general mongodb test", () => {
+    it("mongodb should work", async () => {
+        const postEvent = await Database.EVENTS.create(
+            publicEventValidator.parse(TESTER_EVENT)
+        );
+        const getEvent = await Database.EVENTS.findOne({ eventId: "b" });
+        expect(postEvent.toObject()).toEqual(getEvent?.toObject());
+    });
+});
```
Comment (AydanPirani) [inline src/app.test.ts:38]: move this into event test lwk

File: testing/jest.db-setup.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+    const uri = `${mongod.getUri()}`;
+    if (mongoose.connections.length > 0) {
+        await mongoose.disconnect();
+    }
+    await mongoose.connect(
+        `${uri}?retryWrites=true&w=majority&appName=rp-dev-cluster`
+    );
+    mockConfig(`${uri}?retryWrites=true&w=majority&appName=rp-dev-cluster`);
+});
```
Comment (AydanPirani) [inline testing/jest.db-setup.ts:38]: are we creating the database before each test? seems like a lot of resource consumption

Comment (AydanPirani) [reply testing/jest.db-setup.ts:38]: imo:

beforeAll: spawn mongo inmem
afterAll: destroy mongo inmem

beforeEach: connect to server
afterEach: drop all the databases/remove connection


this way the overhead of spinning up server is only 1x per N tests, as opposed to Nx

File: testing/testingTools.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    return setRole(request(app()).get(url), role);
+}
+
+export function post(url: string, role?: string): request.Test {
+    return setRole(request(app()).post(url), role);
+}
+
+export function put(url: string, role?: string): request.Test {
+    return setRole(request(app()).put(url), role);
+}
```
Comment (AydanPirani) [inline testing/testingTools.ts:50]: dont posts and puts take in request bodies? whats the workaround?

**Review by AydanPirani — CHANGES_REQUESTED** [review]
few changes, then we can merhge

File: src/app.test.ts (hunk 1 above)
Comment (aletya) [reply src/app.test.ts:38]: Even tho it doesn't test the /events endpoints? Cuz this is just testing that the local in memory mongodb server is up.

File: testing/testingTools.ts (hunk 3 above)
Comment (aletya) [reply testing/testingTools.ts:50]: ex:
const response = await post("/events", "admin").send(EVENT)

File: src/app.test.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        const response = await get("/status", undefined).expect(StatusCodes.OK);
+
+        expect(response.text).toBe("API is alive!");
+    });
+});
+
+describe("general mongodb test", () => {
+    it("in-memory mongodb server should work", async () => {
+        const postEvent = await Database.EVENTS.create(
+            publicEventValidator.parse(TESTER_EVENT)
```
Comment (AydanPirani) [inline src/app.test.ts:33]: you don't need to parse it
```suggestion
            TESTER_EVENT
```

File: src/app.test.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        const getEvent = await Database.EVENTS.findOne({ eventId: "b" });
+        expect(getEvent).not.toBeNull();
+        expect(postEvent!.toObject()).toMatchObject(getEvent!.toObject());
+    });
+});
+
+describe("jest setup test", () => {
+    it("creating an object in the db", async () => {
+        await Database.EVENTS.create(publicEventValidator.parse(TESTER_EVENT));
```
Comment (AydanPirani) [inline src/app.test.ts:45]: same thing, no parse

File: testing/jest.db-setup.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import mongoose from "mongoose";
+
+function mockConfig(dbUrl: string) {
+    jest.mock("../src/config.ts", () => {
+        const actual = jest.requireActual("../src/config.ts") as typeof Config;
+
+        const newConfig: typeof Config.default = {
+            ...actual.default,
+            DATABASE_HOST: dbUrl,
+            ENV: "TESTING",
```
Comment (AydanPirani) [inline testing/jest.db-setup.ts:13]: use the actual enum value here, for ENV

File: testing/testingTools.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+export function patch(url: string, role?: string): request.Test {
+    return setRole(request(app()).patch(url), role);
+}
+
+export function del(url: string, role?: string): request.Test {
```
Comment (AydanPirani) [inline testing/testingTools.ts:56]: delete
```suggestion
export function delete(url: string, role?: string): request.Test {
```

**Review by AydanPirani — APPROVED** [review]
few changes, good afterwards.

File: testing/testingTools.ts (hunk 7 above)
Comment (aletya) [reply testing/testingTools.ts:56]: delete is a reserved word sadly :(

**Pull request merged**